tests/e2e/ctl_v3_auth_cluster_test.go

```diff
@@ -0,0 +1,113 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestAuthCluster(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := &e2e.EtcdProcessClusterConfig{
+		ClusterSize:   1,
+		InitialToken:  "new",
+		SnapshotCount: 2,
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	epc, err := e2e.NewEtcdProcessCluster(ctx, t, cfg)
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+	defer func() {
+		if err := epc.Close(); err != nil {
+			t.Fatalf("could not close test cluster (%v)", err)
+		}
+	}()
+
+	epcClient := epc.Client()
+	createUsers(ctx, t, epcClient)
+
+	if err := epcClient.AuthEnable(ctx); err != nil {
+		t.Fatalf("could not enable Auth: (%v)", err)
+	}
+
+	numKeys := 10
+	testUserClientOpts := e2e.WithAuth("test", "testPassword")
+	rootUserClientOpts := e2e.WithAuth("root", "rootPassword")
+
+	// write more than SnapshotCount keys to single leader to make sure snapshot is created
+	for i := 0; i <= numKeys; i++ {
+		if err := epc.Client(testUserClientOpts).Put(ctx, fmt.Sprintf("/test/%d", i), "test", config.PutOptions{}); err != nil {
+			t.Fatalf("failed to Put (%v)", err)
+		}
+	}
+
+	// start second process
+	if err := epc.StartNewProc(ctx, t, rootUserClientOpts); err != nil {
+		t.Fatalf("could not start second etcd process (%v)", err)
+	}
+
+	// make sure writes to both endpoints are successful
+	for _, endpoint := range epc.EndpointsV3() {
+		if err := epc.Client(testUserClientOpts, e2e.WithEndpoints([]string{endpoint})).Put(ctx, "/test/key", endpoint, config.PutOptions{}); err != nil {
+			t.Fatalf("failed to write to Put to %q (%v)", endpoint, err)
+		}
+		numKeys++
+	}
+
+	// verify cluster state is consistent
+	hashKvs, err := epc.Client(rootUserClientOpts).HashKV(ctx, int64(numKeys))
+	if err != nil {
+		t.Fatalf("could not HashKV (%v)", err)
+	}
+
+	revisionSet := false
+	var revision int64
+	var hash uint32
+	for _, v := range hashKvs {
+		if !revisionSet {
+			revision = v.Header.GetRevision()
+			hash = v.Hash
+			revisionSet = true
+			continue
+		}
+		if revision != v.Header.GetRevision() || hash != v.Hash {
+			t.Errorf("Inconsistent revisions found:")
+			for _, v := range hashKvs {
+				t.Errorf("%+v", v)
+			}
+			t.Fail()
+		}
+	}
```

Comment:
```suggestion
	for i:=0; i<3; i++ {
		hashKvs, err := epc.Client(rootUserClientOpts).HashKV(ctx, 0)
		if err != nil {
			t.Fatalf("failed to get HashKV: %v", err)
		}
		
		assert.Equal(t, 2, len(hashKvs)
		if hashKvs[0].Header.Revision != hashKvs[1].Header.Revision {
			time.Sleep(100 * time.Millisecond)
			continue
		}
		assert.Equal(t, hashKvs[0].Hash, hashKvs[1].Hash)
		return
	}
	
	t.Error("failed to check the two members' hashkv")
```

Reply:
ugh, looks like something else is broken either in e2e env or in hashkv implementation.
Code you suggested fails sometimes passes checks and sometimes fails in `assert.Equal(t, 2, len(hashKvs)` since duplicate entries for same endpoint are returned in hashkv response:
```
/Users/oleg/src/etcd-veshij/etcd/bin/etcdctl (/Users/oleg/src/etcd-veshij/etcd/bin/etcdctl_--user=root:rootPassword_--endpoints=http://localhost:20000,http://localhost:20005_endpoint_hashkv_--endpoints_http://localhost:20000,http://localhost:20005_--rev_0_-w_json) (91694): [{"Endpoint":"http://localhost:20000","HashKV":{"header":{"cluster_id":238453183653593855,"member_id":14578408409545168728,"revision":14,"raft_term":3},"hash":410100985,"compact_revision":-1}},{"Endpoint":"http://localhost:20005","HashKV":{"header":{"cluster_id":238453183653593855,"member_id":325135822923097936,"revision":14,"raft_term":3},"hash":410100985,"compact_revision":-1}},{"Endpoint":"http://localhost:20000","HashKV":{"header":{"cluster_id":238453183653593855,"member_id":14578408409545168728,"revision":14,"raft_term":3},"hash":410100985,"compact_revision":-1}},{"Endpoint":"http://localhost:20005","HashKV":{"header":{"cluster_id":238453183653593855,"member_id":325135822923097936,"revision":14,"raft_term":3},"hash":410100985,"compact_revision":-1}}]
```

Reply:
Feel free to have a deep dive into this:
1. check the value of len([endpoints](https://github.com/etcd-io/etcd/blob/c3fa71b96050be22f00ba0aa8f5d6be38838c557/tests/framework/e2e/etcdctl.go#L35)) before requesting the hashKV;

We can easily tell whether it's the client(test case) issue or (etcdctl or etcdserver issue)

Reply:
It's also OK to investigate this in a separate PR.

Reply:
There is a bug in test HashKV impl - we shouldn't set endpoints there since it's already set in cmdArgs().
Not sure why it's triggered only in some cases, but that's out of scope.
I'll cut a PR to fix.
https://github.com/etcd-io/etcd/blob/9bc4a63a419cd316f6920e37ec551b87d1c7c3a7/tests/framework/e2e/etcdctl.go#L358

Reply:
https://github.com/etcd-io/etcd/pull/14637